### PR TITLE
Fixed #32975 -- Fixed admin system check for inlines with foreign keys to proxy models.

### PR DESCRIPTION
**Pull request by taliraj**
https://code.djangoproject.com/ticket/32975

### CODE REVIEW

File: tests/inline_formsets/tests.py [hunk 1]
```diff
@@ -133,6 +133,17 @@ def test_fk_name_not_foreign_key_field_from_child(self):
...
+    def test_foreign_key_on_proxy_model_with_fk_name(self):
+        # Regression test for #32975.
+        inlineformset_factory(SpecialJournalist, Article, fk_name='journalist', fields="__all__")
+
+
+    def test_foreign_key_on_proxy_model_without_fk_name(self):
+        # Regression test for #32975.
+        inlineformset_factory(SpecialJournalist, Article, fields="__all__")
+
+
```
Comment (felixxm) [inline tests/inline_formsets/tests.py:146]: I would move regression tests to the `modeladmin.test_checks.FkNameCheckTests` with inline classes, e.g.
```python
    def test_proxy_model_parent(self):
        class Parent(Model):
            pass

        class ProxyChild(Parent):
            class Meta:
                proxy = True

        class ProxyProxyChild(ProxyChild):
            class Meta:      
                proxy = True 
                             
        class Related(Model):
            proxy_child = ForeignKey(ProxyChild, on_delete=CASCADE)
                                   
        class InlineFkName(admin.TabularInline):
            model = Related
            fk_name = 'proxy_child'
        
        class InlineNoFkName(admin.TabularInline):
            model = Related
        
        class ProxyProxyChildAdminFkName(admin.ModelAdmin):
            inlines = [InlineFkName, InlineNoFkName]
                                    
        self.assertIsValid(ProxyProxyChildAdminFkName, ProxyProxyChild)

```

File: django/forms/models.py [hunk 2]
```diff
@@ -1009,9 +1009,15 @@ def _get_foreign_key(parent_model, model, fk_name=None, can_fail=False):
         fks_to_parent = [f for f in opts.fields if f.name == fk_name]
         if len(fks_to_parent) == 1:
             fk = fks_to_parent[0]
-            if not isinstance(fk, ForeignKey) or \
-                    (fk.remote_field.model != parent_model and
-                     fk.remote_field.model not in parent_model._meta.get_parent_list()):
+            remote_field_is_proxy = fk.remote_field and fk.remote_field.model._meta.proxy
```
Comment (felixxm) [inline django/forms/models.py:1012]: In both cases we're checking `fk.remote_field.model._meta.proxy` when `fk` is a `ForeignKey` so there is no need to check `fk.remote_field`.

File: django/forms/models.py [hunk 3]
```diff
@@ -1009,9 +1009,15 @@ def _get_foreign_key(parent_model, model, fk_name=None, can_fail=False):
...
+            if not isinstance(fk, ForeignKey) or (
+                not remote_field_is_proxy and
+                fk.remote_field.model != parent_model and
+                fk.remote_field.model not in parent_model._meta.get_parent_list()
+            ) or (
+                remote_field_is_proxy and
+                fk.remote_field.model._meta.proxy_for_model not in parent_model._meta.get_parent_list()
+            ):
```
Comment (felixxm) [inline django/forms/models.py:1020]: In both cases we're checking `fk.remote_field.model._meta.proxy` when `fk` is a `ForeignKey` so there is no need to check `fk.remote_field`. I would also add `parent_list` variable and small comments:
```suggestion
            parent_list = parent_model._meta.get_parent_list()
            if not isinstance(fk, ForeignKey) or (
                # ForeignKey to proxy models.
                fk.remote_field.model._meta.proxy and 
                fk.remote_field.model._meta.proxy_for_model not in parent_list
            ) or (
                # ForeignKey to concrete models. 
                not fk.remote_field.model._meta.proxy and
                fk.remote_field.model != parent_model and
                fk.remote_field.model not in parent_list
            ):
```

File: django/forms/models.py [hunk 4]
```diff
@@ -1025,7 +1031,10 @@ def _get_foreign_key(parent_model, model, fk_name=None, can_fail=False):
             f for f in opts.fields
             if isinstance(f, ForeignKey) and (
                 f.remote_field.model == parent_model or
-                f.remote_field.model in parent_model._meta.get_parent_list()
+                f.remote_field.model in parent_model._meta.get_parent_list() or (
+                    f.remote_field.model and
+                    f.remote_field.model._meta.proxy_for_model in parent_model._meta.get_parent_list()
+                )
```
Comment (felixxm) [inline django/forms/models.py:1037]: We're checking `get_parent_list()` in a loop so I would cache it in the variable:
```python
        parent_list = parent_model._meta.get_parent_list()
        fks_to_parent = [
            f for f in opts.fields
            if isinstance(f, ForeignKey) and (
                f.remote_field.model == parent_model or
                f.remote_field.model in parent_list or (                       
                    f.remote_field.model._meta.proxy and
                    f.remote_field.model._meta.proxy_for_model in parent_list
                )   
            )   
        ]
```

**Review by felixxm — COMMENTED** [review]
@taliraj Thanks for this patch :+1: I left small comments.

**Comment by taliraj** [comment]
@felixxm thanks for the review. Added your improvements to the PR.

**Comment by felixxm** [comment]
@taliraj Thanks :+1: Welcome aboard :boat:

**Pull request merged**